src/librustc_ast_lowering/path.rs

```diff
@@ -366,22 +366,27 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
         param_mode: ParamMode,
         mut itctx: ImplTraitContext<'_, 'hir>,
     ) -> (GenericArgsCtor<'hir>, bool) {
-        let &AngleBracketedArgs { ref args, ref constraints, .. } = data;
-        let has_non_lt_args = args.iter().any(|arg| match arg {
-            ast::GenericArg::Lifetime(_) => false,
-            ast::GenericArg::Type(_) => true,
-            ast::GenericArg::Const(_) => true,
+        let has_non_lt_args = data.args.iter().any(|arg| match arg {
+            AngleBracketedArg::Arg(ast::GenericArg::Lifetime(_)) => false,
+            AngleBracketedArg::Arg(ast::GenericArg::Type(_) | ast::GenericArg::Const(_))
+            | AngleBracketedArg::Constraint(_) => true,
         });
-        (
-            GenericArgsCtor {
-                args: args.iter().map(|a| self.lower_generic_arg(a, itctx.reborrow())).collect(),
-                bindings: self.arena.alloc_from_iter(
-                    constraints.iter().map(|b| self.lower_assoc_ty_constraint(b, itctx.reborrow())),
-                ),
-                parenthesized: false,
-            },
-            !has_non_lt_args && param_mode == ParamMode::Optional,
-        )
+        let args = data
+            .args
+            .iter()
+            .filter_map(|arg| match arg {
+                AngleBracketedArg::Arg(arg) => Some(self.lower_generic_arg(arg, itctx.reborrow())),
+                AngleBracketedArg::Constraint(_) => None,
+            })
+            .collect();
+        let bindings = self.arena.alloc_from_iter(data.args.iter().filter_map(|arg| match arg {
```

Comment:
Instead of `filter_map`ing twice, can we use `partition`?

Reply:
In this case we want to partition into `(A, B)` where `A != B` rather than `(A, A)` which you get with `.partition`. We also need to extract the actual thing inside the `Arg` and `Constraint`, so it doesn't seem like `.partition` is helpful here. Although I could use https://docs.rs/itertools/0.9.0/itertools/trait.Itertools.html#method.partition_map here which seems like the optimal solution.

Reply:
Ah, `partion_map` is what I was really thinking of, yes. Unfortunately we don't seem to have it in the standard library yet… Never mind, then.

Reply:
`rustc_mir` has itertools as a dependency, so I can add it to lowering as well :)

Reply:
Hmm actually, I think I'm going to keep it as is after all... the arena allocation here doesn't make the interaction with partitioning a clear win.